File: src/main/java/no/ntnu/webdev/webproject7/product/ProductController.java
package no.ntnu.webdev.webproject7.product;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("products")
public class ProductController {
    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @GetMapping("")
    public ResponseEntity<List<Product>> getAll() {
        return new ResponseEntity<>(this.productService.getAllProducts(), HttpStatus.OK);
    }

    @PostMapping("")
    public ResponseEntity<String> addOne(@RequestBody Product product) {
        return this.productService.addProduct(product)
                ? new ResponseEntity<>(HttpStatus.OK)
                : new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }

    @PutMapping("")
    public ResponseEntity<String> update(@RequestBody Product product) {
        return this.productService.updateProduct(product)
                ? new ResponseEntity<>(HttpStatus.OK)
                : new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<String> delete(@PathVariable String id) {
        return this.productService.deleteProduct(id)
                ? new ResponseEntity<>(HttpStatus.OK)
                : new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }
}

File: src/main/java/no/ntnu/webdev/webproject7/product/ProductController.kt
package no.ntnu.webdev.webproject7.product

import no.ntnu.webdev.webproject7.crud.CrudController
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("products")
class ProductController(productService: ProductService) : CrudController<Product, ProductId>(productService) {

    @GetMapping("/categories")
    fun getCategories(): ResponseEntity<MutableSet<Category>> {
        return ResponseEntity(getCategoryMap(), HttpStatus.OK);
    }

    private fun getCategoryMap(): MutableSet<Category> {
        return Category.values().toHashSet();
    }
}
